cmd/fluxctl/save_cmd.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	yamls := bufio.NewScanner(bytes.NewReader(config))
+	yamls.Split(splitYAMLDocument)
+
+	for yamls.Scan() {
+		var object saveObject
+		// Most unwanted fields are ignored at this point
+		if err := yaml.Unmarshal(yamls.Bytes(), &object); err != nil {
+			return errors.Wrap(err, "unmarshalling exported yaml")
+		}
+
+		// Filter out remaining unwanted keys from unstructured fields
+		// e.g. .Spec and .Metadata.Annotations
+		filterObject(object)
+
+		if err := saveYAML(object); err != nil {
+			return errors.Wrap(err, "saving yaml object")
+		}
+	}
+
+	if yamls.Err() != nil {
+		return errors.Wrap(yamls.Err(), "splitting exported yaml")
+	}
+
+	return nil
+}
+
+// Remove any data that should not be version controlled
+func filterObject(object saveObject) {
+	delete(object.Metadata.Annotations, "deployment.kubernetes.io/revision")
+	delete(object.Metadata.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
+	deleteNested(object.Spec, "template", "metadata", "creationTimestamp")
+	deleteEmptyMapValues(object.Spec)
+}
+
+// Recurse through nested maps to remove a key
+func deleteNested(m map[interface{}]interface{}, keys ...string) {
+	switch len(keys) {
+	case 0:
+		return
+	case 1:
+		delete(m, keys[0])
+	default:
+		if v, ok := m[keys[0]].(map[interface{}]interface{}); ok {
+			deleteNested(v, keys[1:]...)
+		}
+	}
+}
+
+// Recursively delete map keys with empty values
+func deleteEmptyMapValues(i interface{}) bool {
+	switch i := i.(type) {
+	case map[interface{}]interface{}:
+		if len(i) == 0 {
+			return true
+		} else {
+			for k, v := range i {
+				if deleteEmptyMapValues(v) {
+					delete(i, k)
+				}
+			}
+		}
+	case []interface{}:
+		if len(i) == 0 {
+			return true
+		} else {
+			for _, e := range i {
+				deleteEmptyMapValues(e)
+			}
+		}
+	case nil:
+		return true
+	}
+	return false
+}
+
+// Save YAML to directory structure
+func saveYAML(object saveObject) error {
+	var path string
+	if object.Kind == "Namespace" {
+		path = fmt.Sprintf("%s-ns.yaml", object.Metadata.Name)
+	} else {
+		dir := object.Metadata.Namespace
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+			return errors.Wrap(err, "making directory for namespace")
+		}
+
+		shortKind := abbreviateKind(object.Kind)
+		path = fmt.Sprintf("%s/%s-%s.yaml", dir, object.Metadata.Name, shortKind)
+	}
+
+	fmt.Printf("Saving %s '%s' to %s\n", object.Kind, object.Metadata.Name, path)
+
+	file, err := os.Create(path)
+	if err != nil {
+		return errors.Wrap(err, "creating yaml file")
+	}
+	defer file.Close()
+
+	buf, err := yaml.Marshal(object)
+	if err != nil {
+		return errors.Wrap(err, "marshalling yaml")
+	}
+
+	if _, err := file.Write(buf); err != nil {
+		return errors.Wrap(err, "writing yaml file")
+	}
+
+	return nil
+}
+
+func abbreviateKind(kind string) string {
+	switch kind {
+	case "Service":
+		return "svc"
+	case "ReplicationController":
+		return "rc"
+	case "Deployment":
+		return "dep"
+	default:
```

Comment:
I think we have
```
  case "DaemonSet": return "ds"
  case "ConfigMap": return "config"
  case "Secret": return "secret"
```
too. (Could lowercase in the default case to cover that last, and be consistent otherwise.)